src/nsrt_learning/segmentation.py

```diff
@@ -46,6 +51,47 @@ def _switch_fn(t: int) -> bool:
     return _segment_with_switch_function(trajectory, _switch_fn)
 
 
+def _segment_with_oracle(trajectory: GroundAtomTrajectory) -> List[Segment]:
+    """Segment a trajectory using oracle NSRTs.
+
+    If options are known, just uses _segment_with_option_changes().
+
+    Otherwise, starting at the beginning of the trajectory, keeps track of
+    which oracle ground NSRTs are applicable. When any of them have their
+    effects achieved, that marks the switch point between segments.
+    """
+    traj, all_atoms = trajectory
+    if not traj.actions or traj.actions[0].has_option():
+        return _segment_with_option_changes(trajectory)
+    env = get_or_create_env(CFG.env)
+    gt_nsrts = get_gt_nsrts(env.predicates, env.options)
+    ground_nsrts = {
+        ground_nsrt
+        for nsrt in gt_nsrts
+        for ground_nsrt in utils.all_ground_nsrts(nsrt, list(traj.states[0]))
+    }
+    atoms = all_atoms[0]
+    current_nsrts = list(utils.get_applicable_operators(ground_nsrts, atoms))
+
+    def _switch_fn(t: int) -> bool:
+        nonlocal current_nsrts  # update at each switch point
+        atoms = all_atoms[t + 1]
+        # Check if any of the current NSRT effects hold.
+        for ground_nsrt in current_nsrts:
+            # Check effects.
+            if not ground_nsrt.add_effects.issubset(atoms) or (
```

Comment:
highly confusing...can we just use utils.apply_operator somehow?

at the very least, split this into two if statements, one after the other.